File: cc/path-sum/path-sum.cc
#include <bits/stdc++.h>
using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
typedef TreeNode *node;
class Solution {
public:
  /*
  // elegant version
  bool hasPathSum(node root, int sum) {
    if(!root) return false;
    if(!root->left && !root->right && sum==root->val) return true;
    return hasPathSum(root->left, sum-root->val) || hasPathSum(root->right,sum-root->val);
  }
  */

  bool dfs(node root, int sum, int target) {
    sum += root->val;
    if(!root->left && !root->right) return sum == target;
    else if(root->left && !root->right) return dfs(root->left, sum, target);
    else if(!root->left && root->right) return dfs(root->right, sum, target);
    else return (dfs(root->left, sum, target) || dfs(root->right, sum, target));
  }
  bool hasPathSum(node root, int sum) {
    if(!root) return false;
    return dfs(root, 0, sum);
  }
};

//const int N = ((1<<5)-1);
const int N = 1;
int main(int argc, const char *argv[])
{
  int A[N+1] = {0};
  A[1] = 1;
  //A[2] = 2;
  //A[4] = 3;
  //A[8] = 4;
  //A[16] = 5;
  node nodes[N+1] = { 0 };
  for(int i=1; i<N+1; i++) {
    if(A[i]) nodes[i] = new TreeNode(A[i]);
  }
  for(int i=1; i<N+1; i++) {
    if(2*i < N+1 && nodes[i]) nodes[i]->left = nodes[2*i];
    if(2*i+1 < N+1 && nodes[i]) nodes[i]->right = nodes[2*i+1];
  }
  puts("level order");
  int lv = 0;
  while((1<<lv) <= N) ++lv;
  for(int i=0; i<lv; i++) {
    for(int j=0; j<(1<<i) && (1<<i)+j < N+1; j++) 
      if(nodes[(1<<i)+j]) printf("%d ", nodes[(1<<i)+j]->val);
      else printf("# ");
    puts("");
  }
  puts("--------");

  Solution so;
  bool res = so.hasPathSum(nodes[1], 1);
  printf("%d", res); puts("");
  return 0;
    //if(!root) return sum == target;
}

File: rs/src/quests/path_sum.rs
/**
 * [112] Path Sum
 *
 * Given the root of a binary tree and an integer targetSum, return true if the tree has a root-to-leaf path such that adding up all the values along the path equals targetSum.
A leaf is a node with no children.
 
Example 1:
Input: root = [5,4,8,11,null,13,4,7,2,null,null,null,1], targetSum = 22
Output: true
Explanation: The root-to-leaf path with the target sum is shown.

Example 2:
Input: root = [1,2,3], targetSum = 5
Output: false
Explanation: There two root-to-leaf paths in the tree:
(1 --> 2): The sum is 3.
(1 --> 3): The sum is 4.
There is no root-to-leaf path with sum = 5.

Example 3:
Input: root = [], targetSum = 0
Output: false
Explanation: Since the tree is empty, there are no root-to-leaf paths.

 
Constraints:

	The number of nodes in the tree is in the range [0, 5000].
	-1000 <= Node.val <= 1000
	-1000 <= targetSum <= 1000

 */
pub struct Solution {}
use super::super::util::tree::{TreeNode, to_tree};

// submission codes start here

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        if let Some(root) = root {
            let r = root.borrow();
            let (left, right) = (r.left.clone(), r.right.clone());
            if left.is_none() && right.is_none() {
                target_sum == r.val
            } else {
                Self::has_path_sum(left, target_sum - r.val) || Self::has_path_sum(right, target_sum - r.val)
            }
        } else {
            false
        }
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        assert!(Solution::has_path_sum(tree![5,4,8,11,null,13,4,7,2,null,null,null,1], 22));
        assert!(!Solution::has_path_sum(tree![1,2,3], 5));
        assert!(Solution::has_path_sum(tree![], 0));
    }
}
